src/java/htsjdk/samtools/cram/build/CompressionHeaderFactory.java

```diff
@@ -60,128 +50,121 @@
 import java.util.Set;
 import java.util.TreeMap;
 
-public class CompressionHeaderFactory {
-    private static final Charset charset = Charset.forName("US-ASCII");
+/**
+ * Implementation of {@link htsjdk.samtools.cram.build.CramCompression} that mostly relies on GZIP and RANS.
+ */
+public class CompressionHeaderFactory implements CramCompression {
     private static final Log log = Log.getInstance(CompressionHeaderFactory.class);
-    private static final int oqz = ReadTag.nameType3BytesToInt("OQ", 'Z');
-    private static final int bqz = ReadTag.nameType3BytesToInt("BQ", 'Z');
 
-    public CompressionHeader build(final List<CramCompressionRecord> records, final SubstitutionMatrix substitutionMatrix, final boolean sorted) {
+    private final Map<Integer, EncodingDetails> bestEncodings = new HashMap<Integer, EncodingDetails>();
+
+    /**
+     * Decides on compression methods to use for the given records.
+     *
+     * @param records            the data to be compressed
+     * @param substitutionMatrix a matrix of base substitution frequencies, can be null, in which case it is re-calculated.
+     * @param sorted             if true the records are assumed to be sorted by alignment
+     *                           position
+     * @return {@link htsjdk.samtools.cram.structure.CompressionHeader} object describing the encoding chosen for the data
+     */
+    @Override
+    public CompressionHeader buildCompressionHeader(final List<CramCompressionRecord> records,
+                                                    final SubstitutionMatrix substitutionMatrix, final boolean sorted) {
         final CompressionHeader header = new CompressionHeader();
         header.externalIds = new ArrayList<Integer>();
         int exCounter = 0;
 
         final int baseID = exCounter++;
```

Comment:
Why doesn't baseID follow the same pattern as the others?  Please refactor to use CompressionHeader.addExternalCompressor as well.